### PR TITLE
Add listing creation

### PR DESCRIPTION
**Pull request by hpdouglass**
Added listing creation functionality and created a basic webform to test that functionality.

### CODE REVIEW

File: listing_create.py [hunk 1]
```diff
@@ -1,4 +1,45 @@
...
+            return str(e), 400
+    else:
+        # Otherwise, request method was GET so load the form.
+        return render_template('listing_create.html')
+
+def listing_object_from_params(form, files):
+    # Check that all required attributes are present in the request.
+    reqs = ["title", "desc", "price", "author", "password"]
+    for req in reqs:
+        if not form.__contains__(req):
```
Comment (bxw201) [inline listing_create.py:25]: `__contains__` isn't usually called directly as a method, rather it overrides the `in` operator. this shoule be
```python
if not(req in form):
```

File: listing_create.py [hunk 2]
```diff
@@ -1,4 +1,45 @@
...
+    salt = bcrypt.gensalt()
+    hash = bcrypt.hashpw(form.get("password").encode('utf8'), salt)
+    
+    # Return database-ready listing. Errors may still arise in this section.
+    return {
+        "timestamp": int(time.time()),
+        "title": form.get("title"),
+        "description": form.get("desc"),
+        "price": round(float(form.get("price")), 2),
+        "images": [], # TODO: Handle images.
```
Comment (bxw201) [inline listing_create.py:38]: I can handle images once everything else with listing creation is set

File: listing_create.py (hunk 1 above)
Comment (hpdouglass) [reply listing_create.py]: Added a new commit below to fix this, that's a byproduct of me not being as familiar as I would like to be with python. Thanks for the info!

File: listing_create.py [hunk 3]
```diff
@@ -1,4 +1,45 @@
...
 def listing_create():
-    return render_template('newlisting.html')
+    if request.method == "POST":
+        # If request method is POST process submitted form data.
+        try:
+            # If no errors, add listing to database and redirect to new listing page.
+            listing = listing_object_from_params(request.form, request.files)
+            result = listings_collection.insert_one(listing)
+            return redirect(url_for("listing", id=str(result.inserted_id)))
+        except Exception as e:
```
Comment (bxw201) [inline listing_create.py:14]: There are two different types of Exceptions that might be thrown here: one due to incorrect user input raised in `listing_object_from_params` and one due to a possible connection problem with mongodb in `Collection.insert_one`. Should the behavior depending on which one is thrown be different? If the user filled out the form wrong, maybe we could have them keep trying to fill out the form but with the correct fields? Or at least if it errors due to mongodb, that should be a 5xx level error.

Comment (hpdouglass) [reply listing_create.py:14]: Good catch, I thought errors would have been caught earlier when connecting to the database but there could definitely be cases where there are database issues after the connection has already been established. I'll use TypeErrors for validation (return HTTP 400) and then just catch general Exceptions with HTTP 500.

**Pull request merged**